hub/api/pytorch.py

```diff
@@ -0,0 +1,241 @@
+    shm = SharedMemory(create=True, size=len(chunk_bytes), name=chunk)
+    shm.buf[:] = chunk_bytes
+    shm.close()
+    return
+
+
+def remove_shared_memory_from_resource_tracker():
+    """Helper function to fix bug in Python SharedMemory
+
+    Monkey-patch multiprocessing.resource_tracker so SharedMemory won't be tracked
+    More details at: https://bugs.python.org/issue38119
+    """
+
+    def fix_register(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.register(self, name, rtype)
+
+    def fix_unregister(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.unregister(self, name, rtype)
+
+    resource_tracker.register = fix_register
+    resource_tracker.unregister = fix_unregister
+    if "shared_memory" in resource_tracker._CLEANUP_FUNCS:
+        del resource_tracker._CLEANUP_FUNCS["shared_memory"]
+
+
+def _to_pytorch(dataset, transform=None, workers=1):
+    try:
+        import torch
+    except ModuleNotFoundError:
+        raise ModuleNotInstalledException("torch")
+    global torch
+    global _hub_storage_provider  # global to pass to processes, as not possible to serialize it and send as argument
+
+    # TODO, apparently boto3.client isn't safe for multiprocessing https://github.com/boto/boto3/pull/2848/files
+    # could it be working here as we're only reading data?
+    _hub_storage_provider = dataset.provider
+    return TorchDataset(dataset, transform, workers)
+
+
+class TorchDataset:
+    def __init__(self, dataset, transform=None, workers=1):
+        self.dataset = dataset  # TODO disable the memory cache
+        self.transform = transform
+        self.workers = workers
+        self.storage = self.dataset.provider
+        self.map = ProcessPool(nodes=workers).map
+
+        # contains index_map for each Tensor
+        self.all_index_maps = self.load_index_maps()
+
+        # contains meta for each Tensor
+        self.all_meta = self.load_meta()
+
+        # corresponding to each Tensor, stores index-value map. here value is the actual array at the index for the key
+        # this essentially acts as our in memory prefetch cache
+        self.all_index_value_maps = defaultdict(dict)
+
+        # for each Tensor, stores the last index that was prefetched in the prefetch cache
+        self.last_index_map = {}
+
+        # list of all the final samples generated after prefetching and transforming, currently present in memory
+        self.processed_samples = None
+        self.processed_range = slice(-1, -1)  # start and end range of processed_samples
+
+        # keeps track of names of all chunks across tensors whose data is currently prefetched
+        self.all_chunk_sets = {}
+
+        # keeps a pointer to all shared memory objects so that they don't get dereferenced between calls to getitem
+        self.all_shared_memory = []
```

Comment:
for all of these variables can you preface them with `_`? like `self.all_shared_memory`?

Reply:
added `_` to helper function names, but I don't think it helps if we have all internal attributes prefixed with it.